### PR TITLE
#1837: Properly escape project nameofwork on output

### PR DESCRIPTION
**Pull request by chrismiceli**
This was just a global search for the project name. I picked up author name when I ran into it as well. I noticed some other fields appear vulnerable in some contexts, however I did not get to all of those. I've deployed this to my sandbox, but don't know any accounts with permission to create a project:
https://www.pgdp.org/~chrismiceli/c/

### CODE REVIEW

**Comment by srjfoo** [comment]
A couple of screenshots. First, to show an existing project on the test site with HTML tags in both nameofwork and authorsname:
image

From your sandbox, looks good:
image

**Comment by srjfoo** [comment]
I haven't gotten through all of the code changes yet, but one of the changes I saw distracted me and prompted me to look at selected round and pool pages. See https://www.pgdp.org/~chrismiceli/c/project.php?id=projectID44dcfc8b0b855 (P2 page), https://www.pgdp.org/c/project.php?id=projectID564a504488cb6 (PPV page)  and https://www.pgdp.org/~chrismiceli/c/project.php?id=projectID45c417362c7a5 (SR page) in your sandbox.

I may be wrong, but I think that at least one of the relevant pieces of code is `function show_project_listing` in `pinc/showavailablebooks.inc`.

**Comment by chrismiceli** [comment]
> 
> 
> I haven't gotten through all of the code changes yet, but one of the changes I saw distracted me and prompted me to look at selected round and pool pages. See https://www.pgdp.org/~chrismiceli/c/project.php?id=projectID44dcfc8b0b855 (P2 page), https://www.pgdp.org/c/project.php?id=projectID564a504488cb6 (PPV page) and https://www.pgdp.org/~chrismiceli/c/project.php?id=projectID45c417362c7a5 (SR page) in your sandbox.
> 
> I may be wrong, but I think that at least one of the relevant pieces of code is `function show_project_listing` in `pinc/showavailablebooks.inc`.

I think I'm missing something. Which portions distracted you? I'm sure I missed some things.

**Comment by srjfoo** [comment]
> > I haven't gotten through all of the code changes yet, but one of the changes I saw distracted me and prompted me to look at selected round and pool pages. See https://www.pgdp.org/~chrismiceli/c/project.php?id=projectID44dcfc8b0b855 (P2 page), https://www.pgdp.org/c/project.php?id=projectID564a504488cb6 (PPV page) and https://www.pgdp.org/~chrismiceli/c/project.php?id=projectID45c417362c7a5 (SR page) in your sandbox.
> > I may be wrong, but I think that at least one of the relevant pieces of code is `function show_project_listing` in `pinc/showavailablebooks.inc`.
> 
> I think I'm missing something. Which portions distracted you? I'm sure I missed some things.

Apologies -- it wasn't anything that I saw that I thought was wrong, just that until I was reading through the code, I hadn't thought about looking at the round and pool page listings. And yes, good chance we'll find some other places.

**Comment by chrismiceli** [comment]
@cpeel, the php linter would let me know if I referenced html_safe without ever referencing misc.inc right?

**Comment by cpeel** [comment]
> @cpeel, the php linter would let me know if I referenced html_safe without ever referencing misc.inc right?

Negative. The linter just checks syntax, it does not check run-time dependencies.

File: tools/project_manager/editproject.php [hunk 1]
```diff
@@ -1031,7 +1031,7 @@ function preview()
         echo "<br>$a<br>$b";
         echo "</th></tr>\n";
         echo "<tr><td colspan='2'>";
-        echo $comments;
+        echo html_safe($comments);
```
Comment (cpeel) [inline tools/project_manager/editproject.php]: Alas, HTML is valid in project comments, so this and other outputs of them shouldn't be escaped.

File: tools/project_manager/page_compare.php [hunk 2]
```diff
@@ -43,7 +43,7 @@ function render()
         output_header("$title: $sub_title", NO_STATSBAR);
 
         echo "<h1>$title</h1>\n";
-        echo "<h2>$sub_title</h2>\n";
+        echo "<h2>".html_safe($sub_title)."</h2>\n";
```
Comment (cpeel) [inline tools/project_manager/page_compare.php]: Please put spaces around concatenation (here and elsewhere you've added them), eg:
```
echo "<h2>" . html_safe($sub_title) . "</h2>\n";
```

Comment (chrismiceli) [reply tools/project_manager/page_compare.php]: Yeah I mimicked what I saw somewhere else. I'll update all my changed string concatenations and verify runtime dependencies are met.

Comment (chrismiceli) [reply tools/project_manager/page_compare.php]: Updated

File: tools/project_manager/editproject.php (hunk 1 above)
Comment (chrismiceli) [reply tools/project_manager/editproject.php]: updated

File: pinc/ProjectSearchResults.inc [hunk 3]
```diff
@@ -8,7 +8,7 @@ include_once($relPath.'wordcheck_engine.inc');
 include_once($relPath.'ProjectSearchResultsConfig.inc');
 include_once($relPath.'forum_interface.inc'); // get_topic_details() & get_url_to_view_topic()
 include_once($relPath.'misc.inc'); // array_get()
-include_once($relPath.'pg.inc'); // get_pg_catalog_url_for_etext
+include_once($relPath.'pg.inc'); // get_pg_catalog_url_for_etext, html_safe()
```
Comment (cpeel) [inline pinc/ProjectSearchResults.inc]: `html_safe()` is in `misc.inc` (the line above) not `pg.inc`

File: tools/project_manager/automodify.php [hunk 4]
```diff
@@ -89,7 +89,7 @@ function ensure_project_blurb( $project )
         echo "\n";
         echo "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n";
         echo "projectid  = {$project['projectid']}\n";
-        echo "nameofwork = \"{$project['nameofwork']}\"\n";
+        echo "nameofwork = \"".html_safe($project['nameofwork'])."\"\n";
```
Comment (cpeel) [inline tools/project_manager/automodify.php]: Spaces around concatenator, please.

File: tools/project_manager/editproject.php [hunk 5]
```diff
@@ -1016,12 +1016,12 @@ function preview()
...
         echo "<table class='basic'>";
-        echo "<tr><th>", _("Title"), "</th><td>$this->nameofwork</td></tr>\n";
-        echo "<tr><th>", _("Author"), "</th><td>$this->authorsname</td></tr>\n";
+        echo "<tr><th>", _("Title"), "</th><td>", html_safe($this->nameofwork), "</td></tr>\n";
+        echo "<tr><th>", _("Author"), "</th><td>", html_safe($this->authorsname), "</td></tr>\n";
         if (user_is_a_sitemanager())
         {
             // SAs are the only ones who can change this.
-            echo "<tr><th>", _("Project Manager"), "</th><td>$this->projectmanager</td></tr>\n";
+            echo "<tr><th>", _("Project Manager"), "</th><td>", html_safe($this->projectmanager), "</td></tr>\n";
```
Comment (cpeel) [inline tools/project_manager/editproject.php]: You shouldn't need to escape this, these must be valid usernames which can't contain HTML.

File: tools/proofers/md_available.php [hunk 6]
```diff
@@ -42,8 +42,8 @@
            $numpages = Project_getNumPages( $projectid );
 
       echo "<tr>";
-      echo "<td align='right'><a href = \"md_phase1.php?projectid=$projectid\">$name</a></td>\n";
-      echo "<td align='right'>$author</td>\n";
+      echo "<td align='right'><a href = \"md_phase1.php?projectid=$projectid\">".html_safe($name)."</a></td>\n";
```
Comment (cpeel) [inline tools/proofers/md_available.php]: Space around concatenator, please.

File: tools/proofers/md_available.php [hunk 7]
```diff
@@ -86,8 +86,8 @@
            $availpages = Project_getNumPagesInState( $projectid, 'avail_md_second');
 
       echo "<tr>";
-      echo "<td align='right'><a href = \"md_phase2.php?projectid=$projectid\">$name</a></td>\n";
-      echo "<td align='right'>$author</td>\n";
+      echo "<td align='right'><a href = \"md_phase2.php?projectid=$projectid\">".html_safe($name)."</a></td>\n";
```
Comment (cpeel) [inline tools/proofers/md_available.php]: Spaces, please.

File: tools/site_admin/proj_approvals.php [hunk 8]
```diff
@@ -71,8 +71,8 @@
 
         echo "
             <tr>
-            <td class='right-align'><a href='$code_url/project.php?id=$projectid'>$name</a></td>
-            <td class='right-align'>$author</td>
+            <td class='right-align'><a href='$code_url/project.php?id=$projectid'>".html_safe($name)."</a></td>
```
Comment (cpeel) [inline tools/site_admin/proj_approvals.php]: Spaces, please.

File: tools/project_manager/automodify.php (hunk 4 above)
Comment (chrismiceli) [reply tools/project_manager/automodify.php]: whoops, thanks.

File: tools/proofers/md_available.php (hunk 7 above)
Comment (cpeel) [reply tools/proofers/md_available.php]: I think this is the only remaining thing and it's good to merge.

**Pull request merged**